### ISSUE TITLE
webpack output is no longer colored in git bash (again)

### ISSUE THREAD

**Issue by meriton42** [angular/angular-cli#17924]
# 🐞 Bug report

### Command (mark with an `x`)

- [x] serve

... and probably others

### Is this a regression?

Yes. I had already reported and fixed this for angular 6 (#11833, #11834), and the fix was still working as of angular 8.0.

### Reproduction

in git bash, run `ng serve`.

Expected: the console output is colored (since the terminal supports that)
Actual: the console output doesn't use colors

## 🌍 Your Environment

<pre><code>
     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 9.1.8
Node: 12.18.0
OS: win32 ia32

Angular: 9.1.11
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.8
@angular-devkit/build-angular     0.901.8
@angular-devkit/build-optimizer   0.901.8
@angular-devkit/build-webpack     0.901.8
@angular-devkit/core              9.1.8
@angular-devkit/schematics        9.1.8
@angular/cli                      9.1.8
@ngtools/webpack                  9.1.8
@schematics/angular               9.1.8
@schematics/update                0.901.8
rxjs                              6.5.5
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

## Analysis

It seems that this regression was introduced by c4313f55e4412dba50a8d09c4daa4a20642b8e77. Specifically, color support checking is now done in https://github.com/angular/angular-cli/blob/1cd51e64c7331a9cc48d59e8fd6df74d3c6c48f3/packages/angular/cli/utilities/color.ts#L11-L14

Alas, when node is run within a MINGW terminal (as installed by git for windows), `process.stdout` is a `Socket`, causing `instanceof WriteStream` to return false, and all use of colors by the angular CLI to be disabled, even though the MINGW terminal supports colors just fine ...

An easy fix is:

```javascript
// Typings do not contain the function call (added in Node.js v9.9.0)
export const supportsColor =
  process.stdout instanceof WriteStream
  && ((process.stdout as unknown) as { getColorDepth(): number }).getColorDepth() > 1 
  || process.env.MSYSTEM; // MINGW terminals support color even though nodejs mistakes them for a Socket
```

If desired, I can submit a PR to that effect.

Alternatively, you might prefer to delegate to angular_devkit/core/src/terminal/caps.ts instead, whose vendored implementation of `supportsColor` still contains my earlier fix. However, not knowing your policy for dependencies across npm packages I don't feel comfortable submitting a PR to that effect.

**Comment by clydin** [comment]
That solution would break legitimate use cases as well.  There's no safe way to determine whether it is a MINGW simulated terminal or a pipe to another program.  The later of which should not have any color and cannot be used with any terminal APIs.  This overall can affect a variety of other functionality as well; prompts, for instance, may act erratically or not at all.

This is unfortunately an issue with MINGW's simulation of a POSIX terminal and cannot be safely fixed outside of MINGW itself.

**Comment by meriton42** [comment]
Better solutions are of course welcome - but closing this issue as unsolvable seems surprising given that it has previously been solved in two different ways: Early versions of the CLI allowed for overriding color support with `process.env.FORCE_COLOR`, and from angular 6 to 8 we automatically detected MINGW through `process.env.MSYSTEM`. As far as I know, nobody has reported an issue that this was causing them problems. 

Yes, I understand that if somebody uses a MINGW shell to invoke a script that redirects the output of ng build (or similar), the control characters would be annoying. However, this seems a very unusual usecase? I mean, if your build is complicated enough to warrant that kind of automation, chances are you'll be using a CI server - and since that CI server isn't launched through MINGW, it wouldn't be affected by the proposed change?

In summary, I think it is really valuable to be able to see at a glance whether a build is nominal in a terminal, and I think it is a clear step backwards if we drop support for that. 

I'd therefore like this issue to remain open so people can collaborate on a better solution.

**Comment by clydin** [comment]
Git for Windows has an [FAQ section](https://github.com/git-for-windows/git/wiki/FAQ#some-native-console-programs-dont-work-when-run-from-git-bash-how-to-fix-it) dedicated to this particular issue.  It provides some additional details regarding the problem including several solutions.  The solutions are also general and do not require special code to be added to every Node.js application to workaround the limitations.

[WSL](https://docs.microsoft.com/en-us/windows/wsl/) is also an option to investigate as it provides POSIX-like support for Windows via a Microsoft supported extension.

**Comment by meriton42** [comment]
You wouldn't need special code if you used `detectColor` rather than writing your own check. In fact, part of your code base already does that.

I had a look at the solutions from the FAQ:

> Run programs that have problems using the winpty utility. This allows you to keep using the nicer mintty terminal, but can become unwieldy if you need the workaround for many programs.

Indeed. Yeah, I'd like a workaround for *all* invocations of `ng`. I guess I could modify the `ng` shell script yarn installs ... but that's really ugly, isn't it? Or write my own launch script, and retrain myself to write "myng serve" rather than "ng serve"? 

Or I guess I could launch bash itself through winpty, but will that even work? What if I issue a linux command from bash? That would execute in the windows terminal, and likely not work as a result. 

> Modify the shortcut for Git Bash to run bash directly without mintty so it uses the default console host and configure it for "Quick Edit", reasonable size and scroll-back and suitable unicode font. You'll still have to live with the other quirks of console host.

Urgh, please don't make me use the native windows console! While I hear it has been improved some for Windows 10, I'm still on Windows 8 ...

> Install and use ConEmu.

This one appears to work, i.e. it is recognized by Angular CLI as a console, but takes a little time to set up properly. I'm going to be using that one for now.

Still, I think it would be nice if Angular CLI were able to work fully with an out of the box git bash and not require the installation of a 3rd party terminal emulator. 

> WSL

requires Windows 10 as far as I know. I'm still on Windows 8.
